pkg/virt-controller/services/template.go

```diff
@@ -168,6 +169,18 @@ func CPUModelLabelFromCPUModel(vmi *v1.VirtualMachineInstance) (label string, er
 	return
 }
 
+func CPUFeatureLabelsFromCPUFeatures(vmi *v1.VirtualMachineInstance) []string {
+	var labels []string
+	if vmi.Spec.Domain.CPU != nil && vmi.Spec.Domain.CPU.Features != nil {
+		if len(vmi.Spec.Domain.CPU.Features) > 0 {
+			for _, feature := range vmi.Spec.Domain.CPU.Features {
+				labels = append(labels, NFD_CPU_FEATURE_PREFIX+feature.Name)
```

Comment:
all features will make it to the list, policies will be addressed by libvirt.
The nfd plugin will label the nodes with all supported features regardless of policies.
Having that said, in case the nfd plugin will label nodes according to policies as well then we will adapt, but at the current stage this should be sufficient.

Reply:
Does the nfd plugin add __all__ of the cpu flags as labels? all 100+ of them or how does it work?

Reply:
@vladikr  the nfd plugin should add all the supported cpu flags (as labels) for each node in the cluster.
(the pluging is still WIP so we might have/expect some fine tuning on that part)

Reply:
ah. it just sounds very odd to me that we are submitting code that depends on something that doesn't yet exist.
Why not wait until the plugin is implemented?

I also find it a bit weird that the plugin will expose such a large number of labels! It would be much reasonable (to me, at least) to expose a cpu family + certain flags and we will match against that. Similar to what libvirt does.

Reply:
this is a 2 phase submission as we did for cpu families +  this feature is enabled/disabled using a feature gate (as mentioned above). 

for the number of labels exposed, true, this could be large. but then again we will label nodes only with the feature and familys they support and can downsize it further if needed.
(see https://github.com/ksimon1/cpu-model-nfd-plugin)

Reply:
> for the number of labels exposed, true, this could be large. but then again we will label nodes only with the feature and familys they support and can downsize it further if needed.
> (see https://github.com/ksimon1/cpu-model-nfd-plugin)

What's reasonable to me, is that the plugin would expose cpu family `Nehalem` and a cpu feature `spec-ctrl` and when I'm specifically looking for hosts with `spec-ctrl` that would be clear. But if I'm looking for hosts with a cpu feature `ssse3` I wouldn't expect the node to expose it as a separate cpu feature label but rather we should figure out that `ssse3` is part of the `Nehalem` family.

Reply:
> this is a 2 phase submission as we did for cpu families + this feature is enabled/disabled using a feature gate (as mentioned above).

I still find it very hard to undertand why should we keep unfuctioning code that doesn't have it's dependecy merged yet, even if it's behind a feature gate.

Reply:
> What's reasonable to me, is that the plugin would expose cpu family `Nehalem` and a cpu feature `spec-ctrl` and when I'm specifically looking for hosts with `spec-ctrl` that would be clear. But if I'm looking for hosts with a cpu feature `ssse3` I wouldn't expect the node to expose it as a separate cpu feature label but rather we should figure out that `ssse3` is part of the `Nehalem` family.

@vladikr right. And I was expecting that we also cut off older models we do not ever plan to support. E.g. Nehalem would be the very first one, and the extra flags would be as reported by libvirt in addition to the base CPU model name. But I guess that's for the plugin discussion rather than here

Reply:
@michalskrivanek cpu plugin already has functionality for cpu model black list, but it works only with kubernetes version of NFD. For openshift NFD some code changes would be needed

Reply:
@vladikr But node selectors do not work that way. Kubernetes does not know that ssse3 is included in Nehalem. It just treats all those as simple labels.

We need the feature gate for CI and deployments without NFD. NFD is optional, you can easily write your own method or use manual labels. Kubevirt here should be stupid and just copy whathever the user (or higher level management!) asked for. It should not care about what cpu labels or what cpu families are exposed or supported as labels. CNV level handles that and configures kubevirt accordingly.

Reply:
> But I guess that's for the plugin discussion rather than here

@michalskrivanek Exactly. kubevirt should not care.

Reply:
Btw just to be clear, we will limit the amount of exposed families / flags on the node label side. The easiest way is to remove all flags that are always supported by the selected baseline. But I think kubevirt just should not care as this will be implemented externally.

Reply:
Sorry for the delay. I was out last week.
I still think kubevirt can be smarter about it, effectively reducing the number of labels that need to be exposed by NFD - that's just weird to expose it all.
After all, kubevirt can easily know which flag belongs to which family and just schedule according to a family (+ additional flag that is an addition - like `spec-ctrl`)

Reply:
I think the most common use case here will be to add the cpu model to the vm spec and maybe additional flags (or change some existing flags behaviour).
On kubevirt side i don't see too many cases of adding all flags of a family to the vm spec. 
The overload of labels on the node side (assigned by nfd)  might be reduced also by that assumption (and by what @MarSik  mentioned above). 

Well defining the cpu models and features for NFD/Libvirt/scheduler  as discussed [here](https://github.com/kubevirt/kubevirt/issues/1760#issuecomment-442622633) should probably be a further step for this PR.